### PR TITLE
Task8 - Create a ZenML pipeline downloading data from BigQuery in one step in local stack

### LINKED ISSUE #13: Create a ZenML pipeline downloading data from BigQuery in one step in local stack

### PR DESCRIPTION
**Pull request by lukyrasocha**
This PR so far only contains a ZenML pipeline to get data from BQ in one step and trains a simple Logistic Regression on that data (and also evaluates it).

I am not sure if you also intended me to create a Dockerfile and an image, this PR should solve: `Create a ZenML pipeline downloading data from BQ in one step in local stack`, but in a comment in the issue #13 you also mentioned something about images and orchestrators which made me a bit confused and unsure if this is the actual full solution.

Also please let me know if you would prefer me to first ask questions on `slack` when I am unsure about something before creating a PR.

Closes #13

### CODE REVIEW

**Comment by halvgaard** [comment]
I prefer questions like this in PR. Perfectly fine.

Seems like it went pretty smooth. Would love to see it running the pipeline in vertex ai pipelines.

**Comment by lukyrasocha** [comment]
I tried to make it run in Vertex AI pipelines following these [docs](https://docs.zenml.io/features/step-operators) and this [tutorial](https://github.com/zenml-io/zenml/tree/main/examples/step_operator_remote_training) but unfortunately couldn't make it work so far. So I proceeded to try to run the pipeline in Kubeflow pipelines, there I got a bit closer (probably because of the better documentation) but also failed. I will try to debug it and maybe ask on their slack channel

**Comment by lukyrasocha** [comment]
Hey @halvgaard  I managed to make a zenML pipeline run in Vertex AI and included hopefully a thorough description in the readme.

File: Task8/README.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+zenml artifact-store register gcp-store --type=gcp --path=gs://zenml-bucket
+```
+2.3. Create the vertex step operator
+```
+zenml step-operator register <NAME> \
+    --type=vertex \
+    --project=<PROJECT-ID> \
+    --region=<REGION> \
+    --machine_type=<MACHINE-TYPE> \
+    --base_image=<CUSTOM_BASE_IMAGE> #this can be left out if you wish to use zenml's default image
```
Comment (halvgaard) [inline Task8/README.md]: Can you only have one base image for all steps?

File: Task8/README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+1.4. Configure and enable container registry in GCP
+
+This registry will be used by ZenML to push your job images that Vertex will use.
+
+a) [Enable](https://cloud.google.com/container-registry/docs) Container Registry
+b) [Authenticate](https://cloud.google.com/container-registry/docs/advanced-authentication) your local `docker` cli with your GCP container registry 
+
+```
+docker pull busybox
```
Comment (halvgaard) [inline Task8/README.md]: is this image used or just an example?

File: Task8/README.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
+# ZenML pipeline downloading data from BQ in one step and running a zenML pipeline in Vertex AI pipelines
+
+## Notes
+In the folder you will find two python scripts: `pipeline.py` and `run.py`, `pipeline.py` showcases how to download data from BQ in zenML pipeline and 
+`run.py` is used to run another zenML pipeline (mnist) remotely in `Vertex AI` pipelines. The reasons for that is, that so far you can run only pipelines
+that contain python modules installable with `zenml integration install <package>` (list of packages is [here](https://docs.zenml.io/features/integrations)).
+And since to download data from `BigQuery` I used the recommended `pandas_gbq` (not included in zenML integrations) it fails when submitting the job to 
+`Vertex AI` (because zenML defined an image where they install only their integrated packages).
+
+Correction:
```
Comment (halvgaard) [inline Task8/README.md]: Was this problem why it is was difficult to make it work? Was it easy to follow documentation or what was challenging getting this pipeline working?

File: Task8/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+That is true by default, but when you are registering a new step-operator you can define your custom base image `--base_image=<CUSTOM_BASE_IMAGE>`
```
Comment (halvgaard) [inline Task8/README.md]: What if you used the GCP default python package for BQ interaction google-cloud-bigquery. Is that included and would that avoid custom images?

**Review by halvgaard — CHANGES_REQUESTED** [review]
Good job. Could you please put together a max 10 min presentation on this?

**Comment by halvgaard** [comment]
Or maybe create a kaggle notebook on this in stead of the presentation. What do you prefer @lukyrasocha? Both :) ?

File: Task8/README.md (hunk 1 above)
Comment (lukyrasocha) [reply Task8/README.md]: I verified it on their slack channel and yes, you can have only one step operator (with one base image) per stack

File: Task8/README.md (hunk 2 above)
Comment (lukyrasocha) [reply Task8/README.md]: Yes, it is. As I understand it, this image is not used in any way. You push the image, which creates a repository in the Container registry, and `zenml` then uses this repository to push their job images that are then used by `Vertex`.

File: Task8/README.md (hunk 4 above)
Comment (lukyrasocha) [reply Task8/README.md]: I tried it with `from google.cloud import bigquery` and the pipeline runs. So this would avoid custom images.

File: Task8/README.md (hunk 3 above)
Comment (lukyrasocha) [reply Task8/README.md]: Yes I would say the main problems were:
- They have some small bugs in their `examples`, and I couldn't figure out where one error was coming from
- In general I found it hard to debug because sometimes the errors are very long and not that informative
- Their documentation requires some background knowledge, which I don't really have, so I needed to figure out some things with trial and error (but revisiting the documentation for the second time I didn't find it that hard to follow it)

**Comment by lukyrasocha** [comment]
hey @halvgaard, I will do both

**Comment by lukyrasocha** [comment]
hi @halvgaard, I started writing the kaggle notebook but discovered some bugs and errors along the way. So I asked on slack, please have a look at this [slack thread](https://zenml.slack.com/archives/C01FWQ5D0TT/p1652286532072619).
What do you think? Should I do just the presentation?

File: Task8/README.md (hunk 1 above)
Comment (halvgaard) [reply Task8/README.md]: > hi @halvgaard, I started writing the kaggle notebook but discovered some bugs and errors along the way. So I asked on slack, please have a look at this [slack thread](https://zenml.slack.com/archives/C01FWQ5D0TT/p1652286532072619). What do you think? Should I do just the presentation?

Check the response in slack. Maybe put it one of the places he suggests. I think the zenfiles repo has pretty high standards so would be cool and very visible.

File: Task8/README.md (hunk 3 above)
Comment (halvgaard) [reply Task8/README.md]: Can you put this in readme subfolder?

**Comment by halvgaard** [comment]
> hi @halvgaard, I started writing the kaggle notebook but discovered some bugs and errors along the way. So I asked on slack, please have a look at this [slack thread](https://zenml.slack.com/archives/C01FWQ5D0TT/p1652286532072619). What do you think? Should I do just the presentation?

Did you see the suggestion about zenfiles?

**Comment by lukyrasocha** [comment]
Hi @halvgaard , yes, sorry that I didn't respond.  I have currently my exams so I haven't had much time to work this week.

I looked into it and it seems like they have problem specific data science examples there, such as predict customer churn using zenML etc. Do you think my example is fitting for this? Or should I try to find more advanced dataset with more advanced pipeline and then run it on Vertex and then submit it?

Because as I see it my example is a more thorough tutorial on what they already have on step operators in their 'examples' folder.

Summary:
1. I can put what I have here to zenfiles
2. I can make what I have into a more practical use case with a more advanced dataset and put it to zenfiles


Please let me know if this makes sense and what do you think.

Edit: I can also ask on their slack to see if the current thing is fitting to be put on zenfiles and adjust it accordingly

File: Task8/README.md (hunk 3 above)
Comment (lukyrasocha) [reply Task8/README.md]: I moved the readme to the subfolder

**Comment by lukyrasocha** [comment]
The presentation can be viewed here https://docs.google.com/presentation/d/1k5UPsbUSrxQC1ICRH4OCCGq8Wjnl_lc7XrgizUCWTl8/edit?usp=sharing
You can just let me know which day you'd like me to present

Now I will try to adjust this project and contribute to zenfiles in a zenfiles fork.

**Comment by lukyrasocha** [comment]
Hi @halvgaard, I created the pull request in zenfiles. You can check it here: https://github.com/zenml-io/zenfiles/pulls

**Comment by halvgaard** [comment]
Cool :) Nice work. Let me know when it goes through. Also write a post in the slack ML channel to let people know we contributed with this + link to the presentation recordings.

**Comment by lukyrasocha** [comment]
hey @halvgaard, the pull request has been merged and it can be viewed [here](https://github.com/zenml-io/zenfiles/tree/main/time-series-forecast). I also made a post in the slack channel about it

**Review by halvgaard — APPROVED** [review]
